clippy_lints/src/string_patterns.rs

```diff
@@ -0,0 +1,239 @@
+    style,
+    "manual char comparison in string patterns"
+}
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for string methods that receive a single-character
+    /// `str` as an argument, e.g., `_.split("x")`.
+    ///
+    /// ### Why is this bad?
+    /// While this can make a perf difference on some systems,
+    /// benchmarks have proven inconclusive. But at least using a
+    /// char literal makes it clear that we are looking at a single
+    /// character.
+    ///
+    /// ### Known problems
+    /// Does not catch multi-byte unicode characters. This is by
+    /// design, on many machines, splitting by a non-ascii char is
+    /// actually slower. Please do your own measurements instead of
+    /// relying solely on the results of this lint.
+    ///
+    /// ### Example
+    /// ```rust,ignore
+    /// _.split("x");
+    /// ```
+    ///
+    /// Use instead:
+    /// ```rust,ignore
+    /// _.split('x');
+    /// ```
+    #[clippy::version = "pre 1.29.0"]
+    pub SINGLE_CHAR_PATTERN,
+    pedantic,
+    "using a single-character str where a char could be used, e.g., `_.split(\"x\")`"
+}
+
+declare_lint_pass!(StringPatterns => [MANUAL_PATTERN_CHAR_COMPARISON, SINGLE_CHAR_PATTERN]);
+
+const PATTERN_METHODS: [(&str, usize); 22] = [
+    ("contains", 0),
+    ("starts_with", 0),
+    ("ends_with", 0),
+    ("find", 0),
+    ("rfind", 0),
+    ("split", 0),
+    ("split_inclusive", 0),
+    ("rsplit", 0),
+    ("split_terminator", 0),
+    ("rsplit_terminator", 0),
+    ("splitn", 1),
+    ("rsplitn", 1),
+    ("split_once", 0),
+    ("rsplit_once", 0),
+    ("matches", 0),
+    ("rmatches", 0),
+    ("match_indices", 0),
+    ("rmatch_indices", 0),
+    ("trim_start_matches", 0),
+    ("trim_end_matches", 0),
+    ("replace", 0),
+    ("replacen", 0),
+];
+
+fn check_single_char_pattern_lint(cx: &LateContext<'_>, arg: &Expr<'_>) {
+    let mut applicability = Applicability::MachineApplicable;
+    if let Some(hint) = str_literal_to_char_literal(cx, arg, &mut applicability, true) {
+        span_lint_and_sugg(
+            cx,
+            SINGLE_CHAR_PATTERN,
+            arg.span,
+            "single-character string constant used as pattern",
+            "consider using a `char`",
+            hint,
+            applicability,
+        );
+    }
+}
+
+fn get_char_value(expr: &Expr<'_>) -> Option<String> {
+    match expr.kind {
+        ExprKind::Lit(lit) => match lit.node {
+            LitKind::Char(c) => Some(format!("'{}'", c.escape_default())),
+            _ => None,
+        },
+        ExprKind::Path(QPath::Resolved(_, path)) => {
+            if path.segments.len() == 1 {
+                let segment = &path.segments[0];
+                if segment.args.is_none() {
+                    Some(segment.ident.name.to_string())
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
```

Comment:
A check here for `expr.from_expansion()` would be good so we don't suggest crossing any macro boundaries e.g.

```rust
macro_rules! m {
    ($e:expr) => {
        $e == '?'
    };
}
"".find(|c| m!(c));
```

For determining which exprs we can suggest for you could use [`switch_to_eager_eval`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/eager_or_lazy/fn.switch_to_eager_eval.html)

Reply:
Added a test case for this. However `switch_to_eager_eval` doesn't seems to work : 

- the macro call return `true`,
- `(|c: char| c == '.' || c == ',' || c == '!')` return `true`

We are not able to make the distinction, so I used `expr.span.from_expansion()`

Reply:
You'd need both as they do separate things, the `.from_expansion` rules out macro expanded expressions

The `switch_to_eager_eval` is like an extended version of the current `match`, e.g. it will include `foo.bar` as long as there's no user defined `Deref` going on

Reply:
Nice ! ty done.